functions/src/postActions.js

```diff
@@ -0,0 +1,119 @@
+      throw new functions.https.HttpsError(
+        "not-found",
+        `Provided Post ID: ${postId} does not exist`
+      );
+    }
+    return {postRef, postDocSnap};
+}
+
+/// Add current user to the upvote list of a post.
+exports.upVote = functions.https.onCall(async (data, context) => {
+    const uid = context.auth.uid;
+    const postId = data.postId;
+
+    // Get current user
+    const {_, userDocSnap} = await fetchAndValidateUser(uid);
+    // Get target post
+    const {postRef, postDocSnap} = await fetchAndValidatePost(postId);
+
+    // Verify that user has access to post.
+    verifyUserMeetsTokenRequirements(userDocSnap.data(), postDocSnap.data()); 
+
+    // Add user to upVote list if user is not already in upvote list.
+    const upVotes = postDocSnap.data().upVoteUserIds;
+    if (upVotes.indexOf(uid) === -1) {
+      try {
+        upVotes.push(uid);
+        const downVotes = postDocSnap.data().downVoteUserIds.filter((item) => item !== uid);
+        await postRef.set({ upVoteUserIds: upVotes, downVoteUserIds: downVotes }, { merge: true });
+        return {
+          info: `Up vote for Post ID: ${postId} from User ID: ${uid} success`,
+        };
+      } catch (error) {
+        throw new functions.https.HttpsError(
+          "unknown",
+          `Up vote for Post ID: ${postId} from User ID: ${uid} error: ${error?.message}`
+        );
+      }
+    } else {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        `User ID: ${uid}, has already been added to list of up votes for Post ID: ${postId}`
+      );
+    }
+});
+
+/// Add current user to the downvote list of a post.
+exports.downVote = functions.https.onCall(async (data, context) => {
+  const uid = context.auth.uid;
+  const postId = data.postId;
+
+  // Get current user
+  const {_, userDocSnap} = await fetchAndValidateUser(uid);
+  // Get target post
+  const {postRef, postDocSnap} = await fetchAndValidatePost(postId);
+
+  // Verify that user has access to post.
+  verifyUserMeetsTokenRequirements(userDocSnap.data(), postDocSnap.data()); 
+
+  // Add user to upVote list if user is not already in upvote list.
+  const downVotes = postDocSnap.data().downVoteUserIds;
+  if (downVotes.indexOf(uid) === -1) {
+    try {
+      downVotes.push(uid);
+      const upVotes = postDocSnap.data().upVoteUserIds.filter((item) => item !== uid);
+      await postRef.set({ upVoteUserIds: upVotes, downVoteUserIds: downVotes }, { merge: true });
+      return {
+        info: `Down vote for Post ID: ${postId} from User ID: ${uid} success`,
+      };
+    } catch (error) {
+      throw new functions.https.HttpsError(
+        "unknown",
+        `Down vote vote for Post ID: ${postId} from User ID: ${uid} error: ${error?.message}`
+      );
+    }
+  } else {
+    throw new functions.https.HttpsError(
+      "invalid-argument",
+      `User ID: ${uid}, has already been added to list of down votes for Post ID: ${postId}`
+    );
+  }
```

Comment:
https://firebase.google.com/docs/firestore/manage-data/add-data#update_elements_in_an_array
maybe more efficient
but guessed it doesn t change much
if already exist it is not added

Reply:
also could be done in the front end with a firestore rule
allow update: if (token requirement) && affected keys == vote || likes && (affected keys == like? like[request.auth.uid]==null  : false)
or somth

could have like and dislike fields and rules that you are in only one, the front end do the logic 
but since you already made that no need to change